servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/AbstractTimeoutHttpFilter.java

```diff
@@ -0,0 +1,128 @@
+
+abstract class AbstractTimeoutHttpFilter implements HttpExecutionStrategyInfluencer {
+    /**
+     * Establishes the timeout for a given request.
+     */
+    private final TimeoutFromRequest timeoutForRequest;
+
+    /**
+     * If {@code true} then timeout is for full request/response transaction otherwise only the response metadata must
+     * complete before the timeout.
+     */
+    private final boolean fullRequestResponse;
+
+    @Nullable
+    private final Executor timeoutExecutor;
+
+    AbstractTimeoutHttpFilter(final TimeoutFromRequest timeoutForRequest, final boolean fullRequestResponse) {
+        this.timeoutForRequest = requireNonNull(timeoutForRequest, "timeoutForRequest");
+        this.fullRequestResponse = fullRequestResponse;
+        this.timeoutExecutor = null;
+    }
+
+    AbstractTimeoutHttpFilter(final TimeoutFromRequest timeoutForRequest, final boolean fullRequestResponse,
+                              final Executor timeoutExecutor) {
+        this.timeoutForRequest = requireNonNull(timeoutForRequest, "timeoutForRequest");
+        this.fullRequestResponse = fullRequestResponse;
+        this.timeoutExecutor = requireNonNull(timeoutExecutor, "timeoutExecutor");
+    }
+
+    @Override
+    public final HttpExecutionStrategy influenceStrategy(final HttpExecutionStrategy strategy) {
+        return timeoutForRequest.influenceStrategy(strategy);
+    }
+
+    final Single<StreamingHttpResponse> withTimeout(final StreamingHttpRequest request,
+                final Function<StreamingHttpRequest, Single<StreamingHttpResponse>> responseFunction) {
+
+        return Single.defer(() -> {
+            final Duration timeout = timeoutForRequest.apply(request);
+            if (null != timeout && !isPositive(timeout)) {
+                return Single.failed(new TimeoutException("non-positive timeout of " + timeout.toMillis() + "ms"));
+            }
+
+            Single<StreamingHttpResponse> response = responseFunction.apply(request);
+            if (null != timeout) {
+                final Single<StreamingHttpResponse> timeoutResponse = timeoutExecutor == null ?
+                        response.timeout(timeout) : response.timeout(timeout, timeoutExecutor);
+
+                if (fullRequestResponse) {
+                    final long deadline = System.nanoTime() + timeout.toNanos();
+                    response = timeoutResponse.map(resp -> resp.transformMessageBody(body -> defer(() -> {
+                        final Duration remaining = ofNanos(deadline - System.nanoTime());
+                        if (isPositive(remaining)) {
+                            return (timeoutExecutor == null ?
+                                    body.timeoutTerminal(remaining) : body.timeoutTerminal(remaining, timeoutExecutor))
+                                    .subscribeShareContext();
+                        }
+                        return failed(new TimeoutException("timeout after " + timeout.toMillis() + "ms"));
```

Comment:
@bondolo I've removed `subscribeShareContext()` from here because it's not necessary to apply it in case we return failure right away. It's a static async source that does not contain any user code that can modify the context.

Reply:
Understood (now). Thank you for the explanation.